File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/services/storage/informe/InformeFileXml.java
package org.example.services.storage.informe;

import org.example.config.AppConfig;
import org.example.dto.InformeDto;
import org.example.dto.InformesDto;
import org.example.models.Informe;
import org.simpleframework.xml.core.Persister;

import java.io.File;
import java.util.List;

public class InformeFileXml implements InformeStorageService {
    // Singleton
    private static InformeFileXml instance = null;
    public static InformeFileXml getInstance() {
        if (instance == null) {
            instance = new InformeFileXml();
        }
        return instance;
    }

    private final String localFile = AppConfig.getInstance().getAppData() + File.separator + "informes.xml";

    @Override
    public List<Informe> saveAll(List<Informe> elements) {
        File file = new File(localFile);
        if (file.exists() && !file.canWrite() ) return null;
        Persister persister = new Persister();
        try {
            persister.write(
                    new InformesDto(
                            elements.stream().map(Informe::toDto).toList()
                    ), file
            );
            return elements;
        } catch (Exception e) {
            System.out.println("Error al escribir en el fichero xml");
        }
        return null;
    }

    @Override
    public List<Informe> loadAll() {
        File file = new File(localFile);
        if(!file.exists() || !file.canRead()) return null;
        Persister persister = new Persister();
        try {
            InformesDto dto = persister.read(InformesDto.class, file);
            return dto.getInformes().stream().map(InformeDto::toClass).toList();
        } catch (Exception e) {
            System.out.println("Error al leer en el fichero json");
        }
        return null;
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/services/storage/informe/InformeFileXml.kt
package services.storage.informe

import config.AppConfig
import dto.InformesDto
import models.Informe
import mu.KotlinLogging
import org.simpleframework.xml.core.Persister
import java.io.File

private val logger = KotlinLogging.logger {  }

object InformeFileXml: InformeStorageService {
    private val localFile = "${AppConfig.APP_DATA}${File.separator}informe.xml"

    override fun saveAll(elements: List<Informe>): List<Informe> {
        logger.debug { "InformeFileXml ->\tInformeFileXml" }
        val file = File(localFile)
        if (file.exists() && !file.canWrite()) return emptyList()

        val serializer = Persister()
        serializer.write(
            InformesDto(
                elements.map { it.toDto() }
            ),
            file
        )

        return elements
    }

    override fun loadAll(): List<Informe> {
        logger.debug { "InformeFileXml ->\tloadAll" }

        val file = File(localFile)
        if (!file.exists() || !file.canRead()) return emptyList()

        val serializer = Persister()

        return serializer.read(InformesDto::class.java, file).informes.map { it.toClass() }
    }
}